twilight-util/src/signature_validation.rs

```diff
@@ -0,0 +1,184 @@
+            .map(Self)
+            .map_err(KeyError::MalformedKey)
+    }
+    /// Validate a signature for a given message body, timestamp, and signing key.
+    ///
+    /// (This method is a duplicate of [`check_signature`].)
+    ///
+    /// # Errors
+    /// This will fail if the request being validated has an invalid signature, or if it
+    /// was given the wrong key.
+    pub fn verify(
+        &self,
+        signature: &[u8],
+        timestamp: &[u8],
+        body: &[u8],
+    ) -> Result<(), SignatureValidationFailure> {
+        check_signature(signature, timestamp, body, self)
+    }
+}
+
+/// The name of the HTTP header Discord wants us to read for the signature.
+pub const SIGNATURE_HEADER: &str = "x-signature-ed25519";
+/// The name of the HTTP header Discord wants us to read for the signature timestamp.
+pub const TIMESTAMP_HEADER: &str = "x-signature-timestamp";
+
+/// Validates that a signature is valid for a given message body, timestamp, and signing key.
+///
+/// # Errors
+/// This will fail if the request being validated has an invalid signature, or if it
+/// was given the wrong key.
+pub fn check_signature(
+    signature: &[u8],
+    timestamp: &[u8],
+    body: &[u8],
+    key: &Key,
+) -> Result<(), SignatureValidationFailure> {
+    let mut sig_buf = [0; 64];
+    hex::decode_to_slice(signature, &mut sig_buf)
+        .map_err(|e| SignatureValidationFailure::Hex(FromHexError(e)))?;
+    let sig = Signature::from_bytes(&sig_buf);
+
+    let mut buf = Vec::with_capacity(timestamp.len() + body.len());
+    buf.extend_from_slice(timestamp);
+    buf.extend_from_slice(body);
+    match key.0.verify_strict(&buf, &sig) {
+        Ok(()) => Ok(()),
+        Err(e) => Err(SignatureValidationFailure::InvalidSignature(SigError(e))),
+    }
+}
+
+/// Extracting the body of an Interaction failed.
+#[cfg(feature = "signature-validation-extract-interaction")]
+pub enum ExtractFailure {
+    /// The failure was due to the Interaction having an invalid signature.
+    Signature(SignatureValidationFailure),
+    /// The failure was due to the Interaction being incorrect or invalid JSON.
+    Deserialize(serde_json::Error),
+}
+
+#[cfg(feature = "signature-validation-extract-interaction")]
+impl From<SignatureValidationFailure> for ExtractFailure {
+    fn from(value: SignatureValidationFailure) -> Self {
+        Self::Signature(value)
+    }
+}
+
+#[cfg(feature = "signature-validation-extract-interaction")]
+impl From<serde_json::Error> for ExtractFailure {
+    fn from(value: serde_json::Error) -> Self {
+        Self::Deserialize(value)
+    }
+}
+
+/// Validate an Interaction's signature, and deserialize it from JSON.
+///
+/// # Errors
+/// This will fail in the following cases:
+/// - The request has an invalid signature
+/// - The wrong key was given
+/// - Deserialization of the Interaction fails
+#[cfg(feature = "signature-validation-extract-interaction")]
+pub fn extract_interaction(
```

Comment:
I'd like to be convinced why this is useful over users calling `check_signature` and deserializing themselves (optionally using `simd_json`).

Reply:
Personally, I feel like having a function that takes bytes and returns a valid interaction is a design that makes more sense then something returning a Result<(), Error>. It makes it harder to accidentally refactor out the check logic- but I admit this very well may be a contrived problem.

Reply:
I feel like it takes longer for users to understand the function and backing feature and toggle it in `Cargo.toml` than just deserializing the payload themselves. Every public item incurs mental overhead for new users learning Twilight, and Twilight's already complex enough.

Reply:
> I think Discord does a good job validating this already

I think this is a good point. They already do send test requests with invalid signatures, and disable events if you don't correctly reject those requests. So someone doing this would find out pretty quickly.

(I personally wouldn't mind removing this function, but I'd like y'all to come to a consensus so I don't end up deleting it and then adding it back again.)

Reply:
I did say my example was contrived. Removing it seems fine to me.

Reply:
I have removed it for now, but I think we may want to add it back at a later stage, but I see no reason for that decision to block this pr.